backend/src/sse/broadcaster.rs

```diff
@@ -0,0 +1,156 @@
+                this.remove_stale_clients().await;
+            }
+        });
+    }
+
+    /// Removes all non-responsive clients from broadcast list.
+    /// TODO: this is a naive implementation, we should probably use a better data structure for this.
+    ///       Things to consider:
+    ///        - how can we do this without having to iterate over all clients?
+    async fn remove_stale_clients(&self) {
+        let mut guard = self.inner.lock().await;
+
+        let mut ok_maps = HashMap::with_capacity(guard.maps.capacity());
+
+        let old_maps = guard.maps.values();
+
+        for map in old_maps {
+            let mut ok_clients = Vec::with_capacity(map.clients.capacity());
+
+            for client in &map.clients {
+                if client
+                    .sender
+                    .send(sse::Event::Comment("ping".into()))
+                    .await
+                    .is_ok()
+                {
+                    ok_clients.push(client.clone());
+                }
+            }
+
+            if !ok_clients.is_empty() {
+                ok_maps.insert(
+                    map.map_id,
+                    ConnectedMap {
+                        map_id: map.map_id,
+                        clients: ok_clients,
+                    },
+                );
+            }
+        }
+
+        guard.maps = ok_maps;
+    }
+
+    /// Registers client with broadcaster, returning an SSE response body.
+    ///
+    /// # Errors
+    /// * If sender.send() fails for the new client.
+    pub async fn new_client(
+        &self,
+        map_id: i32,
+        user_id: String,
+    ) -> Result<Sse<ChannelStream>, Box<dyn std::error::Error>> {
+        let (sender, channel_stream) = sse::channel(10);
+        let mut guard = self.inner.lock().await;
+
+        let map = guard.maps.entry(map_id).or_insert_with(|| ConnectedMap {
+            map_id,
+            clients: Vec::new(),
+        });
+
+        sender.send(sse::Data::new("connected")).await?;
+
+        map.clients.push(ConnectedClient {
+            id: user_id.clone(),
+            sender,
+        });
+
+        Ok(channel_stream)
+    }
+
+    /// Broadcasts `msg` to all clients.
+    ///
+    /// # Errors
+    /// * If serialization of `msg` fails.
+    pub async fn broadcast<T: serde::Serialize + Send>(
+        &self,
+        map_id: i32,
+        msg: T,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let serialized_data = sse::Data::new_json(msg)?;
+        let guard = self.inner.lock().await;
+
+        if let Some(map) = guard.maps.get(&map_id) {
+            let send_futures = map
+                .clients
+                .iter()
+                .map(|client| client.sender.send(serialized_data.clone()));
+
+            // try to send to all clients, ignoring failures
+            // disconnected clients will get swept up by `remove_stale_clients`
+            let _ = future::join_all(send_futures).await;
```

Comment:
Here both variants should be fine. I personally prefer streams because I have finer control.
```suggestion
let _ = stream::iter(&map.clients)
                .map(|client| client.sender.send(serialized_data.clone()))
                .buffer_unordered(15)
                .collect::<Vec<_>>()
                .await;
```

Reply:
Thanks, I did not know about streams. I will change it later today.